### PR TITLE
Issue 246: Progress report shows erratic p values depending on most recent hint

### LINKED ISSUE #246: Progress report shows erratic p values depending on most recent hint
We talked about this in meeting. we should compute ps for the 3 possible hint states, passing the hint state into the p function each time...

Currently the p code int he tdf defined the hintsyls like below and computes the length to get the intercept. So we can just pass this variable in for each hint syl condition (0,1, and 2), to get the 3 ps instead of generating the 1 randomly.

Then the progress report uses the 0 condition, while the students see items selected from the set of probabilities for all items with 3 probabilities per item

  p.hintsylls = [];
if((p.stimSuccessCount+p.stimFailureCount)<3 && p.syllables>2) {
            var numHintSylls = getRandomInt(3);
            
            for(var index=0;index<numHintSylls;index++){
              p.hintsylls.push(index);
            }
          }
var intercept;
if (p.hintsylls.length==0) {intercept=-1.03873} else if (p.hintsylls.length==1) {intercept=0.19647} else {intercept=0.45817}

### PR DESCRIPTION
**Pull request by JRustyHaner**
Fixes #246 

-Added CF_Hint_Level to history
-CF_Hint_Level is populated by session variable hintLevel
-hintLevel is set to the amount of syllables prefilled in to a cloze

### CODE REVIEW

File: mofacts/server/methods.js [hunk 1]
```diff
@@ -390,26 +390,26 @@ async function setComponentStatesByUserIdTDFIdAndUnitNum(userId, TDFId, componen
...
-            ${probabilityEstimate},${firstSeen},${lastSeen},${trialsSinceLastSeen},${priorCorrect},${priorIncorrect}, \
-            ${priorStudy},${totalPracticeDuration},${outcomeStack}) \
-            RETURNING componentStateId',
-          componentState);
-        } else {
-          resArr.push('not caught error:', e);
+      hintLevel=${hintLevel}, firstSeen=${firstSeen}, lastSeen=${lastSeen}, trialsSinceLastSeen=${trialsSinceLastSeen}, \
+      priorCorrect=${priorCorrect}, priorIncorrect=${priorIncorrect}, \
+      priorStudy=${priorStudy}, totalPracticeDuration=${totalPracticeDuration}, outcomeStack=${outcomeStack} \
+      WHERE userId=${userId} AND TDFId=${TDFId} AND KCId=${KCId} AND componentType=${componentType} \
```
Comment (andrewtackett) [inline mofacts/server/methods.js:396]: In the read functions you treat hintlevel as part of a composite primary key but in this function you don't.  Shouldn't we be querying based on hint level as well here?

**Review by andrewtackett — CHANGES_REQUESTED** [review]
Additionally there are apparently some merge conflicts, please try to pull and resolve those

File: mofacts/client/views/experiment/unitEngine.js [hunk 2]
```diff
@@ -668,14 +683,36 @@ function modelUnitEngine() {
...
+      //Hint Level metrics
+      p.hintLevelProbabilities = Array(4);
+      for(i = 0; i<_.keys(stim).length; i++){
+        let hintLevelProbabilityEstimates = [];
+        try{
+          if(stim[i].hintLevel !== 'undefined'){
+            console.log('Rusty stim[i]', stim[i]);
+            hintLevelProbabilityEstimates.push(stim[i].probabilityEstimate);
+            console.log('Rusty hintLevelProbabilityEstimates ', hintLevelProbabilityEstimates);
+            hintLevelProbability = hintLevelProbabilityEstimates.reduce((a,b) => a + b) / hintLevelProbabilityEstimates.length;
```
[the remark this answers is not in the document]
Comment (JRustyHaner) [reply mofacts/client/views/experiment/unitEngine.js]: at one point I thought it was useful, but it wasn't inside the scope.

**Comment by JRustyHaner** [comment]
@andrewtackett

File: mofacts/server/methods.js [hunk 3]
```diff
@@ -390,26 +390,26 @@ async function setComponentStatesByUserIdTDFIdAndUnitNum(userId, TDFId, componen
...
+        if (e.name == 'QueryResultError') {
+          const componentStateId = await t.one('INSERT INTO componentState(userId,TDFId,KCId,componentType, \
+            probabilityEstimate,hintLevel,firstSeen,lastSeen,trialsSinceLastSeen,priorCorrect,priorIncorrect,priorStudy, \
+            totalPracticeDuration,outcomeStack) VALUES(${userId},${TDFId}, ${KCId}, ${componentType}, \
+            ${probabilityEstimate},${hintLevel},  ${firstSeen},${lastSeen},${trialsSinceLastSeen},${priorCorrect},${priorIncorrect}, \
+            ${priorStudy},${totalPracticeDuration},${outcomeStack}) \
+            RETURNING componentStateId',
+            componentState);
+        } else {
+          esArr.push('not caught error:', e);
```
Comment (andrewtackett) [inline mofacts/server/methods.js:412]: esArr?

**Pull request merged**